src/formula.rs

```diff
@@ -225,6 +247,19 @@ impl Formula {
         Some(Box::new(substitutions))
     }
 
+    /// Unify two sets of formulas
+    /// If successful, this returns the substitutions which needs to be made in
+    /// the second formulas in the tuples in order to match the first formulas.
+    /// In case of failure, returns the index at which the substitution failed.
+    pub fn unify_n(formulas: &[(&Formula, &Formula)]) -> Result<Box<Substitutions>, usize> {
```

Comment:
I think it would be better to expose a function that takes a `&mut Substitutions` and you can do the iteration externally. The `formulas` parameter is baking in a lot of data structure assumptions.

(The error type is also very not-great, I generally expect much better unification errors than just "it failed".)

Reply:
Ok, I've changed the original `unify` function to do that.
For now, I'm adding a generic "UnificationFailed" error which can be later refined.